tta/data/options_data.txt
# Copyright 2010-2026 Free Software Foundation, Inc.
#
# This file is part of GNU Texinfo.
#
# GNU Texinfo is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License,
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# this information is used to setup both the XS options
# and the perl main program and converters options.

# format is:
# option_name   category    default_value   type
# undef is special for default_value, it means undefined.  What undefined
# actually means depends on the variable.  In could mean
#  * unset/false/empty string
#  * set to the default value, when an unset value does not make sense
# By chance, there is no space in default values for now, and therefore
# no need for quoting/escaping.


# Parser only options
IGNORE_SPACE_AFTER_BRACED_COMMAND_NAME  parser 1 integer
# handle cpp like synchronization lines
CPP_LINE_DIRECTIVES                     parser 1 integer
# max number of nested macro calls
MAX_MACRO_CALL_NESTING                  parser 100000 integer
# if set, do not record index entries and ignore index related @-commands
NO_INDEX                                parser 0 integer
# if set, ignore @*macro, @def*index, @alias and @definfoenclose
NO_USER_COMMANDS                        parser 0 integer


# @-commands that can be used multiple time in a document and default
# values.
# also used in util/txicustomvars

allowcodebreaks            multiple_at_command true     char
codequotebacktick          multiple_at_command off      char
codequoteundirected        multiple_at_command off      char
contents                   multiple_at_command 0        integer
deftypefnnewline           multiple_at_command off      char
documentencoding           multiple_at_command utf-8    char

# no @documentlanguage means that the language is not set.
# --document-language
documentlanguage           multiple_at_command undef    char
documentscript             multiple_at_command undef    char

evenfooting                multiple_at_command undef    char
evenheading                multiple_at_command undef    char
everyfooting               multiple_at_command undef    char
everyheading               multiple_at_command undef    char
# is N ems in TeX, 0.4 in.  Can be asis or a number
exampleindent              multiple_at_command 5        char
firstparagraphindent       multiple_at_command none     char
frenchspacing              multiple_at_command off      char
headings                   multiple_at_command on       char
kbdinputstyle              multiple_at_command distinct char
microtype                  multiple_at_command undef    char
oddheading                 multiple_at_command undef    char
oddfooting                 multiple_at_command undef    char
paragraphindent            multiple_at_command 3        integer
shortcontents              multiple_at_command 0        integer
summarycontents            multiple_at_command 0        integer
urefbreakstyle             multiple_at_command after    char
xrefautomaticsectiontitle  multiple_at_command off      char


# @-commands that should be unique.
# also used in util/txicustomvars
afivepaper           unique_at_command undef   char
afourpaper           unique_at_command undef   char
afourlatex           unique_at_command undef   char
afourwide            unique_at_command undef   char
bsixpaper            unique_at_command undef   char
# when passed through a customization variable, documentdescription
# should be already formatted for HTML.  There is no default,
# what is determined to be the title is used if not set.
documentdescription  unique_at_command undef   char
evenfootingmarks     unique_at_command undef   char
evenheadingmarks     unique_at_command undef   char
everyfootingmarks    unique_at_command bottom  char
everyheadingmarks    unique_at_command bottom  char
fonttextsize         unique_at_command 11      integer
# --footnote-style
footnotestyle        unique_at_command end     char
# --no-validate
novalidate           unique_at_command 0       integer
oddfootingmarks      unique_at_command undef   char
oddheadingmarks      unique_at_command undef   char
pagesizes            unique_at_command undef   char
setchapternewpage    unique_at_command on      char
setfilename          unique_at_command undef   char
smallbook            unique_at_command undef   char


# comand-line options that are primarily used to control conversion.
# a value corresponds to defaults that are the same for every output format
# otherwise undef is used
# --split-size
SPLIT_SIZE                converter_cmdline 300000  integer
# --fill-column
FILLCOLUMN                converter_cmdline 72      integer
# --number-sections
NUMBER_SECTIONS           converter_cmdline 1       integer
# --number-footnotes
NUMBER_FOOTNOTES          converter_cmdline 1       integer
# only in HTML
# --transliterate-file-names
TRANSLITERATE_FILE_NAMES  converter_cmdline 0       integer
# --split
SPLIT                     converter_cmdline undef   char
# --headers.  Used to set diverse
# customization options in main program.
# Only directly used in HTML converter
HEADERS                   converter_cmdline 1       integer
# --node-files.  Default depends on SPLIT;  Used in HTML only.
NODE_FILES                converter_cmdline undef   integer
# --verbose
VERBOSE                   converter_cmdline undef   integer
# --output    Sets OUTPUT if non split and not ending by /.
# Setting can be format dependent
OUTFILE                   converter_cmdline undef   char
# --output    Sets SUBDIR if split or ending by /.
# Setting can be format dependent
SUBDIR                    converter_cmdline undef   char
# --disable-encoding/--enable-encoding.
# The option is directly used in
# Info/Plaintext, and used in diverse formats
# for index sorting and plain text output.
ENABLE_ENCODING           converter_cmdline 1       integer


# used in main program, defaults documented in manual
# --macro-expand.  Only for main program
MACRO_EXPAND    program_cmdline undef  bytes
# used in HTML only, called from main program
# --internal-links
INTERNAL_LINKS  program_cmdline undef  bytes
# --error-limit
ERROR_LIMIT     program_cmdline 100    integer
# --force
FORCE           program_cmdline undef  integer
# --no-warn
NO_WARN         program_cmdline undef  integer
# --trace-includes
TRACE_INCLUDES  program_cmdline 0      integer

# following also set in converters
# --headers.  Modified by the format.
FORMAT_MENU     program_cmdline menu   char


# output warnings when node with
# automatic direction and directions in menu are not consistent
# with sectionning, and when node directions are not consistent
# with menu directions.
CHECK_NORMAL_MENU_STRUCTURE  program_customization 1         integer
CHECK_MISSING_MENU_ENTRY     program_customization 1         integer
DUMP_STRUCTURE               program_customization undef     bytes
DUMP_TREE                    program_customization undef     bytes
DUMP_TEXI                    program_customization undef     integer
SHOW_BUILTIN_CSS_RULES       program_customization 0         integer
SORT_ELEMENT_COUNT           program_customization undef     char
SORT_ELEMENT_COUNT_WORDS     program_customization undef     integer
TEXI2DVI                     program_customization texi2dvi  char
TREE_TRANSFORMATIONS         program_customization undef     char


# up node of Top node default value
TOP_NODE_UP                        converter_customization (dir)   char
BASEFILENAME_LENGTH                converter_customization 255-10  integer
CASE_INSENSITIVE_FILENAMES         converter_customization 0       integer
# currently only used in HTML
COPY_IMAGES                        converter_customization 0       integer
DEBUG                              converter_customization 0       integer
DOC_ENCODING_FOR_INPUT_FILE_NAME   converter_customization 1       integer
DOC_ENCODING_FOR_OUTPUT_FILE_NAME  converter_customization 0       integer
# only used in HTML
HANDLER_FATAL_ERROR_LEVEL          converter_customization 100     integer
# only used in HTML
IMAGE_LINK_PREFIX                  converter_customization undef   char
TEST                               converter_customization 0       integer
# for textcontent format
TEXTCONTENT_COMMENT                converter_customization undef   integer
# Not used since the XML format is not generated anymore.
TEXINFO_DTD_VERSION                converter_customization 7.3     char
USE_UNICODE_COLLATION              converter_customization 1       integer

# Some are for all converters, EXTENSION for instance, some for
# some converters, for example CLOSE_QUOTE_SYMBOL and many
# for HTML.
# Defaults are documented in manual and set in the various converters.
# used in util/txicustomvars
AFTER_BODY_OPEN                    converter_customization undef   char
AFTER_SHORT_TOC_LINES              converter_customization undef   char
AFTER_TOC_LINES                    converter_customization undef   char
ASCII_DASHES_AND_QUOTES            converter_customization undef   integer
ASCII_GLYPH                        converter_customization undef   integer
ASCII_PUNCTUATION                  converter_customization undef   integer
AUTO_MENU_DESCRIPTION_ALIGN_COLUMN converter_customization undef   integer
AUTO_MENU_MAX_WIDTH                converter_customization undef   integer
BEFORE_SHORT_TOC_LINES             converter_customization undef   char
BEFORE_TOC_LINES                   converter_customization undef   char
BIG_RULE                           converter_customization undef   char
BODY_ELEMENT_ATTRIBUTES            converter_customization undef   char
# for LaTeX
CLASS_BEGIN_USEPACKAGE             converter_customization undef   char
COPIABLE_LINKS                     converter_customization undef   integer
CHAPTER_HEADER_LEVEL               converter_customization undef   integer
CHECK_HTMLXREF                     converter_customization undef   integer
CLOSE_DOUBLE_QUOTE_SYMBOL          converter_customization undef   char
CLOSE_QUOTE_SYMBOL                 converter_customization undef   char
COLLATION_LANGUAGE                 converter_customization undef   char
COMMAND_LINE_ENCODING              converter_customization undef   char
INDENTED_BLOCK_COMMANDS_IN_TABLE   converter_customization undef   integer
CONTENTS_OUTPUT_LOCATION           converter_customization undef   char
CONVERT_TO_LATEX_IN_MATH           converter_customization undef   integer
DATE_IN_HEADER                     converter_customization undef   integer
DEFAULT_RULE                       converter_customization undef   char
DEF_TABLE                          converter_customization undef   integer
DO_ABOUT                           converter_customization undef   integer
# not documented, used for tests
_DOCBOOK_PIECE                     converter_customization undef   integer
DOCTYPE                            converter_customization undef   char
DOCUMENTLANGUAGE_COLLATION         converter_customization undef   integer
# for LaTeX
END_USEPACKAGE                     converter_customization undef   char
# for ext/epub3.pm
EPUB_CREATE_CONTAINER_FILE         converter_customization undef   integer
# for ext/epub3.pm
EPUB_KEEP_CONTAINER_FOLDER         converter_customization undef   integer
EXTENSION                          converter_customization undef   char
EXTERNAL_CROSSREF_EXTENSION        converter_customization undef   char
# This is used like a boolean, but it is set to a char to match SPLIT, such
# that the value can be set as a copy of SPLIT value
EXTERNAL_CROSSREF_SPLIT            converter_customization undef   char
EXTERNAL_DIR                       converter_customization undef   char
EXTRA_HEAD                         converter_customization undef   char
FOOTNOTE_END_HEADER_LEVEL          converter_customization undef   integer
FOOTNOTE_SEPARATE_HEADER_LEVEL     converter_customization undef   integer
HEADER_IN_TABLE                    converter_customization undef   integer
# for ext/highlight_syntax.pm
HIGHLIGHT_SYNTAX                   converter_customization undef   char
# for ext/highlight_syntax.pm
HIGHLIGHT_SYNTAX_DEFAULT_LANGUAGE  converter_customization undef   char
HTML_MATH                          converter_customization undef   char
HTML_ROOT_ELEMENT_ATTRIBUTES       converter_customization undef   char
HTMLXREF_FILE                      converter_customization undef   char
HTMLXREF_MODE                      converter_customization undef   char
ICONS                              converter_customization undef   integer
INDEX_ENTRY_COLON                  converter_customization undef   char
INDEX_SPECIAL_CHARS_WARNING        converter_customization undef   integer
INFO_JS_DIR                        converter_customization undef   char
INFO_SPECIAL_CHARS_QUOTE           converter_customization undef   integer
INFO_SPECIAL_CHARS_WARNING         converter_customization undef   integer
INFO_MATH_IMAGES                   converter_customization undef   integer
IGNORE_REF_TO_TOP_NODE_UP          converter_customization undef   integer
INLINE_CSS_STYLE                   converter_customization undef   integer
INPUT_FILE_NAME_ENCODING           converter_customization undef   char
JS_WEBLABELS                       converter_customization undef   char
JS_WEBLABELS_FILE                  converter_customization undef   char
LATEX_FLOATS_FILE_EXTENSION        converter_customization tfl     char
LATEX_TAGGED_PDF                   converter_customization undef   char
LOCALE_ENCODING                    converter_customization undef   char
L2H_CLEAN                          converter_customization undef   integer
L2H_FILE                           converter_customization undef   char
L2H_HTML_VERSION                   converter_customization undef   char
L2H_L2H                            converter_customization undef   char
L2H_SKIP                           converter_customization undef   integer
L2H_TMP                            converter_customization undef   char
MATHJAX_CONFIGURATION              converter_customization undef   char
MATHJAX_SCRIPT                     converter_customization undef   char
MATHJAX_SOURCE                     converter_customization undef   char
MAX_HEADER_LEVEL                   converter_customization undef   integer
MENU_ENTRY_COLON                   converter_customization undef   char
MENU_SYMBOL                        converter_customization undef   char
MESSAGE_ENCODING                   converter_customization undef   char
MONOLITHIC                         converter_customization undef   integer
NO_CSS                             converter_customization undef   integer
NO_NUMBER_FOOTNOTE_SYMBOL          converter_customization undef   char
NO_CUSTOM_HTML_ATTRIBUTE           converter_customization undef   integer
NODE_NAME_IN_INDEX                 converter_customization undef   integer
NODE_NAME_IN_MENU                  converter_customization undef   integer
NO_TOP_NODE_OUTPUT                 converter_customization undef   integer
OPEN_DOUBLE_QUOTE_SYMBOL           converter_customization undef   char
OPEN_QUOTE_SYMBOL                  converter_customization undef   char
OUTPUT_CHARACTERS                  converter_customization undef   integer
OUTPUT_ENCODING_NAME               converter_customization undef   char
OUTPUT_FILE_NAME_ENCODING          converter_customization undef   char
PACKAGE                            converter_customization undef   char
PACKAGE_AND_VERSION                converter_customization undef   char
PACKAGE_NAME                       converter_customization undef   char
PACKAGE_URL                        converter_customization undef   char
PACKAGE_VERSION                    converter_customization undef   char
PRE_BODY_CLOSE                     converter_customization undef   char
PREFIX                             converter_customization undef   char
PROGRAM                            converter_customization undef   char
PROGRAM_NAME_IN_ABOUT              converter_customization undef   integer
PROGRAM_NAME_IN_FOOTER             converter_customization undef   integer
SECTION_NAME_IN_TITLE              converter_customization undef   integer
SHORT_TOC_LINK_TO_TOC              converter_customization undef   integer
SHOW_TITLE                         converter_customization undef   integer
T4H_LATEX_CONVERSION               converter_customization undef   char
T4H_MATH_CONVERSION                converter_customization undef   char
T4H_TEX_CONVERSION                 converter_customization undef   char
TEXI2HTML                          converter_customization undef   integer
TEXINFO_OUTPUT_FORMAT              converter_customization undef   char
# Not used, the corresponding formats are not generated anymore
TXI_MARKUP_NO_SECTION_EXTENT       converter_customization undef   integer
TOC_LINKS                          converter_customization undef   integer
TOP_FILE                           converter_customization undef   char
TOP_NODE_FILE_TARGET               converter_customization undef   char
TOP_NODE_UP_URL                    converter_customization undef   char
USE_ACCESSKEY                      converter_customization undef   integer
USE_ISO                            converter_customization undef   integer
USE_LINKS                          converter_customization undef   integer
USE_NODES                          converter_customization undef   integer
USE_NODE_DIRECTIONS                converter_customization undef   integer
USE_NUMERIC_ENTITY                 converter_customization undef   integer
USE_SETFILENAME_EXTENSION          converter_customization undef   integer
USE_TITLEPAGE_FOR_TITLE            converter_customization undef   integer
USE_XML_SYNTAX                     converter_customization undef   integer
VERTICAL_HEAD_NAVIGATION           converter_customization undef   integer
WORDS_IN_PAGE                      converter_customization undef   integer
XREF_USE_FLOAT_LABEL               converter_customization undef   integer
XREF_USE_NODE_NAME_ARG             converter_customization undef   integer

# TODO Not documented for now.
# If set, conversion/formatting not in C.
XS_EXTERNAL_CONVERSION             converter_customization undef   integer
XS_EXTERNAL_FORMATTING             converter_customization undef   integer

# used for development tests, not sure that it will stay in the long term
XS_STRXFRM_COLLATION_LOCALE        converter_customization undef   char


tta/perl/Texinfo/Options.pm
# Automatically generated from regenerate_perl_options_info.pl

package Texinfo::Options;

our %array_cmdline_options = (
  'CSS_FILES'                        => undef,
  'CSS_REFS'                         => undef,
  'EXPANDED_FORMATS'                 => undef,
  'INCLUDE_DIRECTORIES'              => undef,
  'TEXINFO_LANGUAGE_DIRECTORIES'     => undef,
);

our %converter_cmdline_options = (
  'SPLIT_SIZE'                       => 300000,
  'FILLCOLUMN'                       => 72,
  'NUMBER_SECTIONS'                  => 1,
  'NUMBER_FOOTNOTES'                 => 1,
  'TRANSLITERATE_FILE_NAMES'         => 0,
  'SPLIT'                            => undef,
  'HEADERS'                          => 1,
  'NODE_FILES'                       => undef,
  'VERBOSE'                          => undef,
  'OUTFILE'                          => undef,
  'SUBDIR'                           => undef,
  'ENABLE_ENCODING'                  => 1,
);

our %converter_customization_options = (
  'TOP_NODE_UP'                      => '(dir)',
  'BASEFILENAME_LENGTH'              => 255-10,
  'CASE_INSENSITIVE_FILENAMES'       => 0,
  'COPY_IMAGES'                      => 0,
  'DEBUG'                            => 0,
  'DOC_ENCODING_FOR_INPUT_FILE_NAME' => 1,
  'DOC_ENCODING_FOR_OUTPUT_FILE_NAME' => 0,
  'HANDLER_FATAL_ERROR_LEVEL'        => 100,
  'IMAGE_LINK_PREFIX'                => undef,
  'TEST'                             => 0,
  'TEXTCONTENT_COMMENT'              => undef,
  'TEXINFO_DTD_VERSION'              => '7.3',
  'USE_UNICODE_COLLATION'            => 1,
  'AFTER_BODY_OPEN'                  => undef,
  'AFTER_SHORT_TOC_LINES'            => undef,
  'AFTER_TOC_LINES'                  => undef,
  'ASCII_DASHES_AND_QUOTES'          => undef,
  'ASCII_GLYPH'                      => undef,
  'ASCII_PUNCTUATION'                => undef,
  'AUTO_MENU_DESCRIPTION_ALIGN_COLUMN' => undef,
  'AUTO_MENU_MAX_WIDTH'              => undef,
  'BEFORE_SHORT_TOC_LINES'           => undef,
  'BEFORE_TOC_LINES'                 => undef,
  'BIG_RULE'                         => undef,
  'BODY_ELEMENT_ATTRIBUTES'          => undef,
  'CLASS_BEGIN_USEPACKAGE'           => undef,
  'COPIABLE_LINKS'                   => undef,
  'CHAPTER_HEADER_LEVEL'             => undef,
  'CHECK_HTMLXREF'                   => undef,
  'CLOSE_DOUBLE_QUOTE_SYMBOL'        => undef,
  'CLOSE_QUOTE_SYMBOL'               => undef,
  'COLLATION_LANGUAGE'               => undef,
  'COMMAND_LINE_ENCODING'            => undef,
  'INDENTED_BLOCK_COMMANDS_IN_TABLE' => undef,
  'CONTENTS_OUTPUT_LOCATION'         => undef,
  'CONVERT_TO_LATEX_IN_MATH'         => undef,
  'DATE_IN_HEADER'                   => undef,
  'DEFAULT_RULE'                     => undef,
  'DEF_TABLE'                        => undef,
  'DO_ABOUT'                         => undef,
  '_DOCBOOK_PIECE'                   => undef,
  'DOCTYPE'                          => undef,
  'DOCUMENTLANGUAGE_COLLATION'       => undef,
  'END_USEPACKAGE'                   => undef,
  'EPUB_CREATE_CONTAINER_FILE'       => undef,
  'EPUB_KEEP_CONTAINER_FOLDER'       => undef,
  'EXTENSION'                        => undef,
  'EXTERNAL_CROSSREF_EXTENSION'      => undef,
  'EXTERNAL_CROSSREF_SPLIT'          => undef,
  'EXTERNAL_DIR'                     => undef,
  'EXTRA_HEAD'                       => undef,
  'FOOTNOTE_END_HEADER_LEVEL'        => undef,
  'FOOTNOTE_SEPARATE_HEADER_LEVEL'   => undef,
  'HEADER_IN_TABLE'                  => undef,
  'HIGHLIGHT_SYNTAX'                 => undef,
  'HIGHLIGHT_SYNTAX_DEFAULT_LANGUAGE' => undef,
  'HTML_MATH'                        => undef,
  'HTML_ROOT_ELEMENT_ATTRIBUTES'     => undef,
  'HTMLXREF_FILE'                    => undef,
  'HTMLXREF_MODE'                    => undef,
  'ICONS'                            => undef,
  'INDEX_ENTRY_COLON'                => undef,
  'INDEX_SPECIAL_CHARS_WARNING'      => undef,
  'INFO_JS_DIR'                      => undef,
  'INFO_SPECIAL_CHARS_QUOTE'         => undef,
  'INFO_SPECIAL_CHARS_WARNING'       => undef,
  'INFO_MATH_IMAGES'                 => undef,
  'IGNORE_REF_TO_TOP_NODE_UP'        => undef,
  'INLINE_CSS_STYLE'                 => undef,
  'INPUT_FILE_NAME_ENCODING'         => undef,
  'JS_WEBLABELS'                     => undef,
  'JS_WEBLABELS_FILE'                => undef,
  'LATEX_FLOATS_FILE_EXTENSION'      => 'tfl',
  'LATEX_TAGGED_PDF'                 => undef,
  'LOCALE_ENCODING'                  => undef,
  'L2H_CLEAN'                        => undef,
  'L2H_FILE'                         => undef,
  'L2H_HTML_VERSION'                 => undef,
  'L2H_L2H'                          => undef,
  'L2H_SKIP'                         => undef,
  'L2H_TMP'                          => undef,
  'MATHJAX_CONFIGURATION'            => undef,
  'MATHJAX_SCRIPT'                   => undef,
  'MATHJAX_SOURCE'                   => undef,
  'MAX_HEADER_LEVEL'                 => undef,
  'MENU_ENTRY_COLON'                 => undef,
  'MENU_SYMBOL'                      => undef,
  'MESSAGE_ENCODING'                 => undef,
  'MONOLITHIC'                       => undef,
  'NO_CSS'                           => undef,
  'NO_NUMBER_FOOTNOTE_SYMBOL'        => undef,
  'NO_CUSTOM_HTML_ATTRIBUTE'         => undef,
  'NODE_NAME_IN_INDEX'               => undef,
  'NODE_NAME_IN_MENU'                => undef,
  'NO_TOP_NODE_OUTPUT'               => undef,
  'OPEN_DOUBLE_QUOTE_SYMBOL'         => undef,
  'OPEN_QUOTE_SYMBOL'                => undef,
  'OUTPUT_CHARACTERS'                => undef,
  'OUTPUT_ENCODING_NAME'             => undef,
  'OUTPUT_FILE_NAME_ENCODING'        => undef,
  'PACKAGE'                          => undef,
  'PACKAGE_AND_VERSION'              => undef,
  'PACKAGE_NAME'                     => undef,
  'PACKAGE_URL'                      => undef,
  'PACKAGE_VERSION'                  => undef,
  'PRE_BODY_CLOSE'                   => undef,
  'PREFIX'                           => undef,
  'PROGRAM'                          => undef,
  'PROGRAM_NAME_IN_ABOUT'            => undef,
  'PROGRAM_NAME_IN_FOOTER'           => undef,
  'SECTION_NAME_IN_TITLE'            => undef,
  'SHORT_TOC_LINK_TO_TOC'            => undef,
  'SHOW_TITLE'                       => undef,
  'T4H_LATEX_CONVERSION'             => undef,
  'T4H_MATH_CONVERSION'              => undef,
  'T4H_TEX_CONVERSION'               => undef,
  'TEXI2HTML'                        => undef,
  'TEXINFO_OUTPUT_FORMAT'            => undef,
  'TXI_MARKUP_NO_SECTION_EXTENT'     => undef,
  'TOC_LINKS'                        => undef,
  'TOP_FILE'                         => undef,
  'TOP_NODE_FILE_TARGET'             => undef,
  'TOP_NODE_UP_URL'                  => undef,
  'USE_ACCESSKEY'                    => undef,
  'USE_ISO'                          => undef,
  'USE_LINKS'                        => undef,
  'USE_NODES'                        => undef,
  'USE_NODE_DIRECTIONS'              => undef,
  'USE_NUMERIC_ENTITY'               => undef,
  'USE_SETFILENAME_EXTENSION'        => undef,
  'USE_TITLEPAGE_FOR_TITLE'          => undef,
  'USE_XML_SYNTAX'                   => undef,
  'VERTICAL_HEAD_NAVIGATION'         => undef,
  'WORDS_IN_PAGE'                    => undef,
  'XREF_USE_FLOAT_LABEL'             => undef,
  'XREF_USE_NODE_NAME_ARG'           => undef,
  'XS_EXTERNAL_CONVERSION'           => undef,
  'XS_EXTERNAL_FORMATTING'           => undef,
  'XS_STRXFRM_COLLATION_LOCALE'      => undef,
  '_INLINE_STYLE_WIDTH'              => undef,
);

our %converter_other_options = (
  'LINKS_DIRECTIONS'                 => undef,
  'TOP_BUTTONS'                      => undef,
  'TOP_FOOTER_BUTTONS'               => undef,
  'SECTION_BUTTONS'                  => undef,
  'CHAPTER_FOOTER_BUTTONS'           => undef,
  'SECTION_FOOTER_BUTTONS'           => undef,
  'NODE_FOOTER_BUTTONS'              => undef,
  'MISC_BUTTONS'                     => undef,
  'CHAPTER_BUTTONS'                  => undef,
  'ACTIVE_ICONS'                     => undef,
  'PASSIVE_ICONS'                    => undef,
);

our %multiple_at_command_options = (
  'allowcodebreaks'                  => 'true',
  'codequotebacktick'                => 'off',
  'codequoteundirected'              => 'off',
  'contents'                         => 0,
  'deftypefnnewline'                 => 'off',
  'documentencoding'                 => 'utf-8',
  'documentlanguage'                 => undef,
  'documentscript'                   => undef,
  'evenfooting'                      => undef,
  'evenheading'                      => undef,
  'everyfooting'                     => undef,
  'everyheading'                     => undef,
  'exampleindent'                    => '5',
  'firstparagraphindent'             => 'none',
  'frenchspacing'                    => 'off',
  'headings'                         => 'on',
  'kbdinputstyle'                    => 'distinct',
  'microtype'                        => undef,
  'oddheading'                       => undef,
  'oddfooting'                       => undef,
  'paragraphindent'                  => 3,
  'shortcontents'                    => 0,
  'summarycontents'                  => 0,
  'urefbreakstyle'                   => 'after',
  'xrefautomaticsectiontitle'        => 'off',
);

our %parser_options = (
  'IGNORE_SPACE_AFTER_BRACED_COMMAND_NAME' => 1,
  'CPP_LINE_DIRECTIVES'              => 1,
  'MAX_MACRO_CALL_NESTING'           => 100000,
  'NO_INDEX'                         => 0,
  'NO_USER_COMMANDS'                 => 0,
);

our %program_cmdline_options = (
  'MACRO_EXPAND'                     => undef,
  'INTERNAL_LINKS'                   => undef,
  'ERROR_LIMIT'                      => 100,
  'FORCE'                            => undef,
  'NO_WARN'                          => undef,
  'TRACE_INCLUDES'                   => 0,
  'FORMAT_MENU'                      => 'menu',
);

our %program_customization_options = (
  'CHECK_NORMAL_MENU_STRUCTURE'      => 1,
  'CHECK_MISSING_MENU_ENTRY'         => 1,
  'DUMP_STRUCTURE'                   => undef,
  'DUMP_TREE'                        => undef,
  'DUMP_TEXI'                        => undef,
  'SHOW_BUILTIN_CSS_RULES'           => 0,
  'SORT_ELEMENT_COUNT'               => undef,
  'SORT_ELEMENT_COUNT_WORDS'         => undef,
  'TEXI2DVI'                         => 'texi2dvi',
  'TREE_TRANSFORMATIONS'             => undef,
);

our %unique_at_command_options = (
  'afivepaper'                       => undef,
  'afourpaper'                       => undef,
  'afourlatex'                       => undef,
  'afourwide'                        => undef,
  'bsixpaper'                        => undef,
  'documentdescription'              => undef,
  'evenfootingmarks'                 => undef,
  'evenheadingmarks'                 => undef,
  'everyfootingmarks'                => 'bottom',
  'everyheadingmarks'                => 'bottom',
  'fonttextsize'                     => 11,
  'footnotestyle'                    => 'end',
  'novalidate'                       => 0,
  'oddfootingmarks'                  => undef,
  'oddheadingmarks'                  => undef,
  'pagesizes'                        => undef,
  'setchapternewpage'                => 'on',
  'setfilename'                      => undef,
  'smallbook'                        => undef,
);



#################################################

my %regular_options_types;

my %common_parser_regular_options_defaults = (
  'documentlanguage'                 => undef,
  'DEBUG'                            => 0,
  'EXPANDED_FORMATS'                 => [],
  'FORMAT_MENU'                      => 'menu',
  'DOC_ENCODING_FOR_INPUT_FILE_NAME' => 1,
  'COMMAND_LINE_ENCODING'            => undef,
  'INCLUDE_DIRECTORIES'              => [ '.' ],
  'INPUT_FILE_NAME_ENCODING'         => undef,
  'LOCALE_ENCODING'                  => undef,
);

$regular_options_types{'common_parser'} = \%common_parser_regular_options_defaults;

my %converter_common_regular_options_defaults = (
  'PACKAGE'                          => 'texinfo',
  'PACKAGE_NAME'                     => 'GNU Texinfo',
  'PACKAGE_AND_VERSION'              => 'texinfo',
  'PACKAGE_VERSION'                  => '',
  'PACKAGE_URL'                      => 'https://www.gnu.org/software/texinfo/',
  'PROGRAM'                          => '',
);

$regular_options_types{'converter_common'} = \%converter_common_regular_options_defaults;

my %converter_defaults_regular_options_defaults = (
);

$regular_options_types{'converter_defaults'} = \%converter_defaults_regular_options_defaults;

my %html_converter_regular_options_defaults = (
  'BIG_RULE'                         => '<hr>',
  'BODY_ELEMENT_ATTRIBUTES'          => undef,
  'CHAPTER_HEADER_LEVEL'             => 2,
  'CLOSE_QUOTE_SYMBOL'               => undef,
  'CONTENTS_OUTPUT_LOCATION'         => 'after_top',
  'CONVERT_TO_LATEX_IN_MATH'         => undef,
  'INDENTED_BLOCK_COMMANDS_IN_TABLE' => 0,
  'CHECK_HTMLXREF'                   => 1,
  'COPIABLE_LINKS'                   => 1,
  'DATE_IN_HEADER'                   => 0,
  'DEFAULT_RULE'                     => '<hr>',
  'DOCTYPE'                          => '<!DOCTYPE html>',
  'DO_ABOUT'                         => 0,
  'OUTPUT_CHARACTERS'                => 0,
  'EXTENSION'                        => 'html',
  'EXTERNAL_CROSSREF_EXTENSION'      => undef,
  'FOOTNOTE_END_HEADER_LEVEL'        => 4,
  'FOOTNOTE_SEPARATE_HEADER_LEVEL'   => 4,
  'FORMAT_MENU'                      => 'sectiontoc',
  'HEADERS'                          => 1,
  'INDEX_ENTRY_COLON'                => '',
  'INLINE_CSS_STYLE'                 => 0,
  'JS_WEBLABELS'                     => 'generate',
  'JS_WEBLABELS_FILE'                => 'js_licenses.html',
  'MAX_HEADER_LEVEL'                 => 4,
  'MENU_ENTRY_COLON'                 => ':',
  'MENU_SYMBOL'                      => undef,
  'MONOLITHIC'                       => 1,
  'NO_CUSTOM_HTML_ATTRIBUTE'         => 0,
  'NO_CSS'                           => 0,
  'NO_NUMBER_FOOTNOTE_SYMBOL'        => '*',
  'NODE_NAME_IN_MENU'                => 1,
  'OPEN_QUOTE_SYMBOL'                => undef,
  'SECTION_NAME_IN_TITLE'            => 0,
  'SHORT_TOC_LINK_TO_TOC'            => 1,
  'SHOW_TITLE'                       => undef,
  'SPLIT'                            => 'node',
  'TOP_FILE'                         => 'index.html',
  'TOP_NODE_FILE_TARGET'             => 'index.html',
  'USE_ACCESSKEY'                    => undef,
  'USE_ISO'                          => 1,
  'USE_LINKS'                        => 1,
  'USE_NODES'                        => 1,
  'USE_NODE_DIRECTIONS'              => undef,
  'USE_TITLEPAGE_FOR_TITLE'          => 1,
  'WORDS_IN_PAGE'                    => 200,
  'XREF_USE_NODE_NAME_ARG'           => undef,
  'XREF_USE_FLOAT_LABEL'             => 0,
  'xrefautomaticsectiontitle'        => 'on',
);

$regular_options_types{'html_converter'} = \%html_converter_regular_options_defaults;

my %plaintexinfo_converter_regular_options_defaults = (
  'EXTENSION'                        => 'texi',
  'OUTFILE'                          => '-',
);

$regular_options_types{'plaintexinfo_converter'} = \%plaintexinfo_converter_regular_options_defaults;

my %plaintext_regular_options_defaults = (
  'ENABLE_ENCODING'                  => 1,
  'ASCII_DASHES_AND_QUOTES'          => 1,
  'ASCII_GLYPH'                      => 0,
  'FORMAT_MENU'                      => 'nomenu',
  'EXTENSION'                        => 'txt',
  'INFO_SPECIAL_CHARS_WARNING'       => undef,
  'OUTFILE'                          => '-',
  'USE_NODES'                        => 1,
);

$regular_options_types{'plaintext'} = \%plaintext_regular_options_defaults;

my %texi2html_regular_options_defaults = (
  'FORMAT_MENU'                      => 'menu',
  'USE_SETFILENAME_EXTENSION'        => 0,
  'footnotestyle'                    => 'separate',
  'CONTENTS_OUTPUT_LOCATION'         => 'separate_element',
  'FORCE'                            => 1,
  'USE_ACCESSKEY'                    => 0,
  'NODE_NAME_IN_MENU'                => 0,
  'SHORT_TOC_LINK_TO_TOC'            => 0,
  'SHOW_TITLE'                       => 1,
  'USE_LINKS'                        => 0,
  'USE_NODES'                        => 0,
  'SPLIT'                            => '',
  'PROGRAM_NAME_IN_FOOTER'           => 1,
  'PROGRAM_NAME_IN_ABOUT'            => 1,
  'HEADER_IN_TABLE'                  => 1,
  'MENU_ENTRY_COLON'                 => '',
  'INDEX_ENTRY_COLON'                => '',
  'DO_ABOUT'                         => undef,
  'CHAPTER_HEADER_LEVEL'             => 1,
  'BIG_RULE'                         => '<hr style="height: 6px;">',
  'FOOTNOTE_END_HEADER_LEVEL'        => 3,
  'FOOTNOTE_SEPARATE_HEADER_LEVEL'   => 1,
);

$regular_options_types{'texi2html'} = \%texi2html_regular_options_defaults;

sub get_regular_options {
  my $input = shift;
  return $regular_options_types{$input}
}

1;
